src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -343,6 +343,140 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
         appendUnusedGroupByColumn(select_query, source_columns);
 }
 
+/// recursive traversal and check for optimizeGroupByFunctionKeys
+struct KeepFunctionMatcher
+{
+    struct Data
+    {
+        std::unordered_set<String> & should_keep;
+        bool & keep;
+    };
+
+    using Visitor = InDepthNodeVisitor<KeepFunctionMatcher, true>;
+
+    static bool needChildVisit(const ASTPtr & node, const ASTPtr &)
+    {
+        return !(node->as<ASTFunction>());
+    }
+
+    static void visit(ASTFunction * function_node, Data & data)
+    {
+        if ((function_node->arguments->children).empty())
+        {
+            data.keep = true;
+            return;
+        }
+
+        if (data.should_keep.find(function_node->getColumnName()) == data.should_keep.end())
+        {
+            Visitor(data).visit(function_node->arguments);
+        }
+    }
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        if (data.keep)
+            return;
+
+        if (!ast)
+            return;
+
+        if (!(ast->as<ASTFunction>() || ast->as<ASTExpressionList>()))
+        {
+            if ((data.should_keep).find(ast->getColumnName()) == (data.should_keep).end())
```

Comment:
It's not correct, We can call `getColumnName()` only for ASTIdentifiers in this context

```
    if (auto * function_node = ast->as<ASTFunction>())
    {
        visit(function_node, data);
    }
    else if (auto * ident = ast->as<ASTIdentifier>())
    {
        visit(ident, data)
    }
```